### ISSUE TITLE
Check DefaultValuesOfCorrectType validation rule operation

### ISSUE THREAD

**Issue by Shane32** [graphql-dotnet/graphql-dotnet#1818]
The `DefaultValuesOfCorrectType` validation rule checks to be sure that required variables do not have a default value set in the query.  See this example in test `no_required_variables_with_default_values`:

```gql
query UnreachableDefaultValues($a: Int! = 3, $b: String! = ""default"") {
  dog { name }
}
```

The above test currently fails the validation rule because a non-null variable is provided with a default value.

However, field arguments are explicitly allowed to be non-nullable with default values in 5.4.2.1.

Should this behavior for variables be changed to match the behavior for fields? The spec seems unclear for non-null variables.

**Comment by sungam3r** [comment]
@benjie Could you please help?

**Comment by benjie** [comment]
[This part of the spec](https://spec.graphql.org/draft/#sec-All-Variable-Usages-are-Allowed.Allowing-optional-variables-when-default-values-exist) explicitly suggests changing `($booleanArg: Boolean = true)` to `($booleanArg: Boolean! = true)` to avoid ambiguity. 

[This part of the spec](https://spec.graphql.org/draft/#sec-Coercing-Variable-Values) allows for coercing values of non-nullable variables where `hasValue` is not true and a `defaultValue` exists.

I can't find anything that would suggest this is not allowed.

TL;DR: non-null variables can have defaults; this is confirmed by the GraphQL reference implementation.

```graphql
query($username: String! = "benjie") {
  userByUsername(username: $username) {
    name
  }
}
```

^ This is valid.

**Comment by sungam3r** [comment]
@benjie Thank you very much. @Shane32 my guess was confirmed. Thanks to your refactoring, we found a bug :)

**Comment by Shane32** [comment]
I'm working on known issues documentation today, but I should be able to fix it this weekend.  If it doesn't make it into 3.0rc, it should be in 3.0 final.
